Mathlib/CategoryTheory/Triangulated/Adjunction.lean

```diff
@@ -0,0 +1,153 @@
+      have eq := congr_arg (adj.homEquiv _ _ ).toFun hα
+      simp only [homEquiv_counit, Functor.id_obj, Equiv.toFun_as_coe, homEquiv_unit,
+        Functor.comp_obj, Functor.map_comp, unit_naturality_assoc, right_triangle_components,
+        comp_id] at eq
+      rw [eq, ← assoc, assoc _ _ f]
+      change _ ≫ (Triangle.mk (G.map T.mor₁) f g).mor₁ ≫ (Triangle.mk (G.map T.mor₁) f g).mor₂ = 0
+      rw [comp_distTriang_mor_zero₁₂ _ mem, comp_zero]
+    have := isIso_of_yoneda_map_bijective (adj.homEquiv _ _ h) (fun Y => by
+      constructor
+      · intro φ₁ φ₂ hφ
+        rw [← cancel_mono (adj.homEquiv _ _ h)]
+        exact hφ
+      · intro φ
+        obtain ⟨ψ, hψ⟩ := Triangle.coyoneda_exact₁ _ mem (φ ≫ G.map T.mor₃ ≫
+          (G.commShiftIso (1 : ℤ)).hom.app T.obj₁) (by
+            dsimp
+            simp only [assoc]
+            rw [← G.commShiftIso_hom_naturality, ← G.map_comp_assoc,
+              comp_distTriang_mor_zero₃₁ _ hT, G.map_zero, zero_comp, comp_zero])
+        dsimp at ψ hψ
+        obtain ⟨α, hα⟩ : ∃ α, α = φ - ψ ≫ (adj.homEquiv _ _) h := ⟨_, rfl⟩
+        have hα₀ : α ≫ G.map T.mor₃ = 0 := by
+          rw [hα, sub_comp, ← cancel_mono ((Functor.commShiftIso G (1 : ℤ)).hom.app T.obj₁),
+            assoc, sub_comp, assoc, assoc, hψ, zero_comp, sub_eq_zero,
+            ← cancel_mono ((Functor.commShiftIso G (1 : ℤ)).inv.app T.obj₁), assoc,
+            assoc, assoc, assoc, h₂', Iso.hom_inv_id_app, comp_id]
+        suffices ∃ (β : Y ⟶ Z), β ≫ (adj.homEquiv _ _) h = α by
+          obtain ⟨β, hβ⟩ := this
+          refine ⟨ψ + β, ?_⟩
+          dsimp
+          rw [add_comp, hβ, hα, add_sub_cancel]
+        obtain ⟨β, hβ⟩ := T.coyoneda_exact₃ hT ((adj.homEquiv _ _).symm α)
+          ((adj.homEquiv _ _).injective (by simpa [homEquiv_unit, homEquiv_counit] using hα₀))
+        refine ⟨adj.homEquiv _ _ β ≫ f, ?_⟩
+        simpa [homEquiv_unit, h₁'] using congr_arg (adj.homEquiv _ _).toFun hβ.symm)
+    refine isomorphic_distinguished _ mem _ (Iso.symm ?_)
+    refine Triangle.isoMk _ _ (Iso.refl _) (Iso.refl _) (asIso (adj.homEquiv Z T.obj₃ h)) ?_ ?_ ?_
+    · dsimp
+      simp
+    · apply (adj.homEquiv _ _).symm.injective
+      dsimp
+      simp only [homEquiv_unit, homEquiv_counit, Functor.map_comp, assoc,
+        counit_naturality, left_triangle_components_assoc, h₁, id_comp]
+    · dsimp
+      rw [Functor.map_id, comp_id, homEquiv_unit, assoc, ← G.map_comp_assoc, ← h₂,
+        Functor.map_comp, Functor.map_comp, assoc, unit_naturality_assoc, assoc,
+        Functor.commShiftIso_hom_naturality, ← adj.shift_unit_app_assoc,
+        ← Functor.map_comp, right_triangle_components, Functor.map_id, comp_id]
+
+end Adjunction
+
+namespace Equivalence
+
+variable (E : C ≌ D) [E.functor.CommShift ℤ] [E.inverse.CommShift ℤ] [E.CommShift ℤ]
+
+/--
+We say that an equivalence of categories `E` is triangulated if both `E.functor` and
+`E.inverse` are triangulated functors.
+-/
+class IsTriangulated : Prop where
```

Comment:
Actually, we should first define `Adjunction.IsTriangulated`:
```lean
class IsTriangulated : Prop where
  commShift : adj.CommShift ℤ := by infer_instance
  leftAdjoint_isTriangulated : F.IsTriangulated := by infer_instance
  rightAdjoint_isTriangulated : G.IsTriangulated := by infer_instance
```
Then, the corresponding definition for equivalences can be defined as an `abbrev` similarly as for `CommShift`.
Then, you may also include instances for the identity adjunction, and the composition of adjunctions (and similarly for `Equivalence.refl` and `Equivalence.trans`).

Reply:
A constructor for this class should also be provided, saying that it suffices to prove that the left adjoint is triangulated. (The symmetric constructor should also be provided, of course, but the second will have to wait.)

The `Adjunction.IsTriangulated` instances for the identity and the composition first require similar instances for `Adjunction.CommShift`: #20337

Reply:
Done.